com.unity.ml-agents/Runtime/SimpleMultiAgentGroup.cs

```diff
@@ -0,0 +1,143 @@
+using System;
+using System.Linq;
+using System.Collections.Generic;
+
+namespace Unity.MLAgents
+{
+    /// <summary>
+    /// A basic class implementation of MultiAgentGroup.
+    /// </summary>
+    internal class SimpleMultiAgentGroup : IMultiAgentGroup, IDisposable
+    {
+        readonly int m_Id = MultiAgentGroupIdCounter.GetGroupId();
+        HashSet<Agent> m_Agents = new HashSet<Agent>();
+
+
+        public void Dispose()
+        {
+            while (m_Agents.Count > 0)
+            {
+                UnregisterAgent(m_Agents.First());
+            }
+        }
+
+        /// <inheritdoc />
+        public virtual void RegisterAgent(Agent agent)
+        {
+            if (!m_Agents.Contains(agent))
+            {
+                agent.SetMultiAgentGroup(this);
+                m_Agents.Add(agent);
+                agent.OnAgentDisabled += UnregisterAgent;
+            }
+        }
+
+        /// <inheritdoc />
+        public virtual void UnregisterAgent(Agent agent)
+        {
+            if (m_Agents.Contains(agent))
+            {
+                agent.SetMultiAgentGroup(null);
+                m_Agents.Remove(agent);
+                agent.OnAgentDisabled -= UnregisterAgent;
+            }
+        }
+
+        /// <inheritdoc />
+        public int GetId()
+        {
+            return m_Id;
+        }
+
+        /// <summary>
+        /// Get list of all agents currently registered to this MultiAgentGroup.
+        /// </summary>
+        /// <returns>
+        /// List of agents registered to the MultiAgentGroup.
+        /// </returns>
+        public HashSet<Agent> GetRegisteredAgents()
```

Comment:
Would be nice to return this as a readonly interface so that the user doesn't mess with it. It looks like `IReadOnlyCollection<Agent>` might be the best bet: https://stackoverflow.com/a/36923075/224264